packages/ripple-keypairs/src/index.ts

```diff
@@ -38,66 +46,79 @@ const secp256k1 = {
-
+    const derived = derivePrivateKey(entropy, options)
     const privateKey =
-      prefix + derivePrivateKey(entropy, options).toString(16, 64).toUpperCase()
+      SECP256K1_PREFIX + bytesToHex(numberToBytesBE(derived, 32))
 
-    const publicKey = bytesToHex(
-      Secp256k1.keyFromPrivate(privateKey.slice(2))
-        .getPublic()
-        .encodeCompressed(),
-    )
+    const publicKey = bytesToHex(nobleSecp256k1.getPublicKey(derived, true))
     return { privateKey, publicKey }
   },
 
-  sign(message, privateKey): string {
-    return bytesToHex(
-      Secp256k1.sign(hash(message), hexToBytes(privateKey), {
-        canonical: true,
-      }).toDER(),
+  sign(message: ByteArray, privateKey: HexString): string {
+    // Some callers pass the privateKey with the prefix, others without.
+    // elliptic.js implementation ignored the prefix, interpreting it as a
+    // leading zero byte. @noble/curves will throw if the key is not exactly
+    // 32 bytes, so we normalize it before passing to the sign method.
+    // TODO: keep back compat like this, or simply always require prefix as
+    // the ed25519 sign method does.
+    assert.ok(
+      (privateKey.length === 66 && privateKey.startsWith(SECP256K1_PREFIX)) ||
+        privateKey.length === 64,
     )
+    const normed = privateKey.length === 66 ? privateKey.slice(2) : privateKey
+    return nobleSecp256k1
+      .sign(hash(message), normed)
+      .toDERHex(true)
+      .toUpperCase()
   },
 
   verify(message, signature, publicKey): boolean {
-    return Secp256k1.verify(hash(message), signature, hexToBytes(publicKey))
+    const decoded = nobleSecp256k1.Signature.fromDER(signature)
+    return nobleSecp256k1.verify(decoded, hash(message), publicKey)
   },
 }
 
 const ed25519 = {
-  deriveKeypair(entropy: Uint8Array): {
+  deriveKeypair(entropy: ByteArray): {
     privateKey: string
     publicKey: string
   } {
     const prefix = 'ED'
     const rawPrivateKey = hash(entropy)
     const privateKey = prefix + bytesToHex(rawPrivateKey)
     const publicKey =
-      prefix + bytesToHex(Ed25519.keyFromSecret(rawPrivateKey).pubBytes())
+      prefix + bytesToHex(nobleEd25519.getPublicKey(rawPrivateKey))
     return { privateKey, publicKey }
   },
 
-  sign(message, privateKey): string {
-    // caution: Ed25519.sign interprets all strings as hex, stripping
-    // any non-hex characters without warning
-    if (!Array.isArray(message)) {
-      throw new Error('message must be array of octets')
-    }
+  sign(message: ByteArray, privateKey: HexString): string {
+    assert.ok(
+      Array.isArray(message) || message instanceof Uint8Array,
+      'message must be array of octets',
+    )
+    assert.ok(
+      privateKey.length === 66,
+      'private key must be 33 bytes including prefix',
+    )
     return bytesToHex(
-      Ed25519.sign(message, hexToBytes(privateKey).slice(1)).toBytes(),
+      nobleEd25519.sign(new Uint8Array(message), privateKey.slice(2)),
```

Comment:
This assert + slice seem odd to me. We're making sure it's 33 bytes, then ignoring the first byte? Should we allow either 33 or 32 bytes instead since that will be the same to the eventual sign behavior?

Reply:
```typescript
function getAlgorithmFromKey(key: HexString): 'ed25519' | 'ecdsa-secp256k1' {
  const bytes = hexToBytes(key)
  return bytes.length === 33 && bytes[0] === 0xed
    ? 'ed25519'
    : 'ecdsa-secp256k1'
}
```
The above snippet is the routing logic to the particular algorithm for given key. If it was a bit stricter, the back-compat wouldn't have been needed. 

I can't see how you could allow untagged 32 byte ED keys without introducing some ambiguity? 

(side note: I can't actually recall offhand, but I don't think the ed25519 signing is exported from this library except via the wrapper with routing, right?)

Reply:
Maybe there's an argument that for consistency sake ONLY properly tagged, fully 33 byte keys should be accepted? i.e. no special treatment for secpk256k1 keys

Reply:
> fully 33 byte keys should be accepted

There's more nuance there.

The valid secp256k1 PUBLIC keys are: 
1. 65 bytes: 0x4 + 32 x-bytes x + 32 y-bytes
2. 33 bytes: 0x2 + 32 x-bytes
3. 33 bytes: 0x3 + 32 x-bytes # y is odd

Valid secp PRIVATE keys:
1. big endian encoded number > zero and < curve order - (uniformity of random bytes aside) - you could encode that as an arbitrary amount of bytes with left padded zeroes 

getAlgorithmFromKey is basically getAlgorithmFromPrivateOrPublicKey

ed25519 scalars and elements/points both encode to 32 bytes

The 0xED prefix is to differentiate them from secp256k1 keys.

Not that an account id, which is a hash of a public key, DOES include 0xED prefix byte.

Reply:
| Curve     | Key Type        | Prefix (Hex) | Length (bytes) | Description                                                      | Detected Algorithm |
|-----------|-----------------|:------------:|:--------------:|------------------------------------------------------------------|-------------------:|
| ed25519   | Private Key     |     0xED     |       33       | prefix + 32 bytes unsigned LE                                    |            ed25519 |
| ed25519   | Public Key      |     0xED     |       33       | prefix + 32 y-bytes                                              |            ed25519 |
| secp256k1 | Public Key (1)  |     0x02     |       33       | prefix + 32 x-bytes                                              |    ecdsa-secp256k1 |
| secp256k1 | Public Key (2)  |     0x03     |       33       | prefix + 32 x-bytes (y is odd)                                   |    ecdsa-secp256k1 |
| secp256k1 | Public Key (3)  |     0x04     |       65       | prefix + 32 x-bytes + 32 y-bytes                                 |    ecdsa-secp256k1 |
| secp256k1 | Private Key (1) |     None     |       32       | 32 bytes unsigned BE ( 0 > n < order)                            |    ecdsa-secp256k1 |
| secp256k1 | Private Key (2) |     0x00     |       33       | prefix (non significant) + 32 bytes unsigned BE ( 0 > n < order) |    ecdsa-secp256k1 |